### PR TITLE
Estimate improvements

### LINKED ISSUE #1209: Estimate improvements
- static dropdown: select year, than q (optional) and custom date as alternative

### PR DESCRIPTION
**Pull request by DenisVorop**
## PR includes
- [x] Feature

## Related issues
Resolve #1209

## QA Instructions, Screenshots, Recordings
  
Снимок экрана 2023-08-24 в 14 44 25
Снимок экрана 2023-08-24 в 14 44 44
Снимок экрана 2023-08-24 в 14 45 23
Снимок экрана 2023-08-24 в 14 45 36

### CODE REVIEW

**Comment by DenisVorop** [comment]
Снимок экрана 2023-08-25 в 13 54 37
Снимок экрана 2023-08-25 в 13 54 44
Снимок экрана 2023-08-25 в 13 54 50
Снимок экрана 2023-08-25 в 13 55 13

File: src/components/EstimateDate.tsx [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    onChange,
+    setReadOnly,
+}) => {
+    const locale = useLocale();
+    const currentDate = currentLocaleDate({ locale });
+    const [fullDate, setFullDate] = useState(currentDate);
+    const ref = useRef(null);
+
+    useClickOutside(ref, () => {
+        if (!fullDate.includes('_')) return;
```
Comment (awinogradov) [inline src/components/EstimateDate.tsx:43]: Very fragile. You base condition on mask type in external library. It's better to check value. You can save it without mask.

File: src/components/EstimateDate.tsx [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        });
+        setFullDate(estimate.date);
+        onChange?.(estimate);
+    }, [onCreateEstimate, currentDate, setReadOnly, onChange]);
+
+    return (
+        <EstimateOption
+            title={option.title}
+            readOnly={readOnly}
+            onClickIcon={onClickIcon}
```
Comment (awinogradov) [inline src/components/EstimateDate.tsx:124]: Any other options to click? Why not onClick?

File: src/components/EstimateOption.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+}) => {
+    const onClick = useCallback(() => {
+        onClickIcon?.();
+    }, [onClickIcon]);
+
+    return (
+        <StyledWrapper>
+            <StyledContent>
+                {nullable(title, (t) => (
+                    <Text weight="regular" size="s">
```
Comment (awinogradov) [inline src/components/EstimateOption.tsx:42]: It's regular by default

File: src/components/EstimateOption.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                        {t}
+                    </Text>
+                ))}
+
+                {nullable(readOnly, () => (
+                    <IconPlusCircleSolid size="xs" color={gray9} onClick={onClick} style={{ cursor: 'pointer' }} />
```
Comment (awinogradov) [inline src/components/EstimateOption.tsx:48]: Do not use inline style. Use styled component instead.

File: src/components/EstimateYear.tsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    }, [onChange, setReadOnly, value?.y]);
+
+    return (
+        <EstimateOption
+            title={option.title}
+            clue={option.clue}
+            readOnly={readOnly}
+            onClickIcon={onClickIcon}
+            renderTrigger={() => (
+                <div style={{ display: 'flex' }} ref={inputRef}>
```
Comment (awinogradov) [inline src/components/EstimateYear.tsx:94]: Do not use inline style

File: src/components/EstimateYear.tsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                    <Input
+                        onChange={onChangeYear}
+                        value={selectedYear}
+                        brick={'right'}
```
Comment (awinogradov) [inline src/components/EstimateYear.tsx:98]: "right"

File: src/components/OutlinePopup.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
+import React, { ComponentProps, ReactNode } from 'react';
+import styled, { css } from 'styled-components';
+import { Popup } from '@taskany/bricks';
+import { radiusM, gapS, gray5 } from '@taskany/colors';
+
+const arrowDirectionStyles = {
```
Comment (awinogradov) [inline src/components/OutlinePopup.tsx:6]: I'm not sure we need this component. Popup has prop to show and hide arrow. But here in Dropdown arrow must be visible.

**Review by awinogradov — CHANGES_REQUESTED** [review]

File: src/components/EstimateQuarter.tsx [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        [locale, onChange, value],
+    );
+
+    const onClick = useCallback(() => {
+        setReadOnly({
+            quarter: false,
+            year: false,
+            date: true,
+        });
+        onChange?.({ y: value?.y || `${new Date().getFullYear()}`, q: null, date: null });
```
Comment (9teen90nine) [inline src/components/EstimateQuarter.tsx:88]: `yearFromDate(new Date)`? Also you probably should just export currentYear from the same utils file as it's used in other files too

**Review by 9teen90nine — APPROVED** [review]

File: src/components/EstimateOption.tsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+export const EstimateOption: React.FC<EstimateOptionProps> = ({
+    title,
+    clue,
+    readOnly,
+    onClick: onClickIcon,
+    renderTrigger,
+}) => {
+    const onClick = useCallback(() => {
```
Comment (awinogradov) [inline src/components/EstimateOption.tsx:38]: But why?) do we really need to wrap onClick?

File: src/components/EstimateOption.tsx [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+`;
+
+const StyledContent = styled.div`
+    display: flex;
+    align-items: end;
+    gap: ${gapS};
+`;
+
+const StyledIconPlusCircleSolid = styled(IconPlusCircleSolid)`
+    cursor: pointer;
```
Comment (awinogradov) [inline src/components/EstimateOption.tsx:20]: This looks like a bug. Icon with onClick must have this style. Please check it, create issue in bricks if not and place comment with issue link here.

File: src/components/EstimatePopup.tsx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+    useClickOutside(triggerRef, (e) => {
+        if (!popupContentRef.current?.contains(e.target as Node)) {
+            setVisible(false);
+            onClose?.();
+        }
+    });
+
+    return (
+        <div>
```
Comment (awinogradov) [inline src/components/EstimatePopup.tsx:90]: I'm not sure you need div here. I think fragment is enough.

File: src/components/EstimatePopup.tsx [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+            <div ref={triggerRef}>{renderTrigger({ onClick: onToggleVisible })}</div>
+
+            <StyledPopup
+                visible={visible}
+                placement={placement}
+                reference={triggerRef}
+                interactive
+                minWidth={160}
+                maxWidth={180}
```
Comment (awinogradov) [inline src/components/EstimatePopup.tsx:108]: If it fixed why values are different?

File: src/components/EstimateQuarter.tsx [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+import { EstimateOption } from './EstimateOption';
+
+const StyledQuarters = styled.div`
+    display: grid;
+    grid-template-columns: repeat(4, 1fr);
+    gap: ${gapXs};
+`;
+
+const StyledCheckableButton = styled(Button)<{ checked?: boolean }>`
+    background-color: ${({ checked }) => checked && gray6};
```
Comment (awinogradov) [inline src/components/EstimateQuarter.tsx:19]: We don't have this style in bricks? Create issue if not and put comment with link here

**Review by awinogradov — CHANGES_REQUESTED** [review]

File: src/components/EstimateQuarter.tsx (hunk 13 above)
Comment (DenisVorop) [reply src/components/EstimateQuarter.tsx:19]: https://github.com/taskany-inc/bricks/pull/418

**Review by awinogradov — APPROVED** [review]

**Pull request merged**